TypeCobol/Compiler/Nodes/Data.cs

```diff
@@ -16,38 +16,72 @@ namespace TypeCobol.Compiler.Nodes {
     public class DataDivision: GenericNode<DataDivisionHeader>, Parent<DataSection> {
 
         public const string NODE_ID = "data-division";
+
+        /// <summary>
+        /// Lists indices for data sections
+        /// </summary>
+        private enum SectionIndex
+        {
+            FileSection,
+            GlobalStorageSection,
+            WorkingStorageSection,
+            LocalStorageSection,
+            LinkageSection
+        }
+        // The 5 (optional) data sections
+        private DataSection[] _sections = new DataSection[5];
+        public FileSection FileSection => (FileSection) _sections[(int)SectionIndex.FileSection];
+        public GlobalStorageSection GlobalStorageSection => (GlobalStorageSection)_sections[(int)SectionIndex.GlobalStorageSection];
+        public WorkingStorageSection WorkingStorageSection => (WorkingStorageSection)_sections[(int)SectionIndex.WorkingStorageSection];
+        public LocalStorageSection LocalStorageSection => (LocalStorageSection)_sections[(int)SectionIndex.LocalStorageSection];
+        public LinkageSection LinkageSection => (LinkageSection)_sections[(int)SectionIndex.LinkageSection];
+
         public DataDivision(DataDivisionHeader header): base(header) { }
         public override string ID { get { return NODE_ID; } }
 
-        public override void Add(Node child, int index = -1) {
-            if (index <= 0) index = WhereShouldIAdd(child.GetType());
+        public override void Add(Node child, int index = -1)
+        {
+            var maxSectionIndex = UpdateSection(child);
+            if (index <= 0) index = WhereShouldIAdd(maxSectionIndex);
+
             base.Add(child,index);
         }
-        private int WhereShouldIAdd(System.Type section) {
-            if (Tools.Reflection.IsTypeOf(section, typeof(FileSection))) return 0;
-            int ifile = -2;
-            int iglobal = -2;
-            int iworking = -2;
-            int ilocal = -2;
-            int ilinkage = -2;
-            int c = 0;
-            foreach(var child in this.Children()) {
-                if (Tools.Reflection.IsTypeOf(child.GetType(), typeof(FileSection))) ifile = c;
-                else
-                if (Tools.Reflection.IsTypeOf(child.GetType(), typeof(GlobalStorageSection))) iglobal = c;
-                else
-                if (Tools.Reflection.IsTypeOf(child.GetType(), typeof(WorkingStorageSection))) iworking = c;
-                else
-                if (Tools.Reflection.IsTypeOf(child.GetType(), typeof(LocalStorageSection))) ilocal = c;
-                else
-                if (Tools.Reflection.IsTypeOf(child.GetType(), typeof(LinkageSection))) ilinkage = c;
-                c++;
+
+        private int UpdateSection(Node node)
+        {
+            int result = -1;
+            if (node is FileSection fileSection)
+            {
+                result = (int)SectionIndex.FileSection;
+                _sections[result] = fileSection;
```

Comment:
The assignment in the array can be factorized at the end and the `if / else if` can be rewritten as switch expression:

```csharp
        private int UpdateSection(Node node)
        {
            int result = node switch
            {
                Nodes.FileSection => (int)SectionIndex.FileSection,
                Nodes.GlobalStorageSection => (int)SectionIndex.GlobalStorageSection,
                Nodes.WorkingStorageSection => (int)SectionIndex.WorkingStorageSection,
                Nodes.LocalStorageSection => (int)SectionIndex.LocalStorageSection,
                Nodes.LinkageSection => (int)SectionIndex.LinkageSection,
                _ => -1
            };

            if (result != -1)
            {
                _sections[result] = (DataSection)node;
            }

            return result;
        }
```

Reply:
Done